ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouterMetrics.java

```diff
@@ -31,42 +37,139 @@
   // @todo: Ensure all metrics here get updated appropriately.
   // @todo: chunk filling rate metrics.
   // @todo: More metrics for the RequestResponse handling (poll, handleResponse etc.)
-  // Rates
+
+  // Operation rate.
   public final Meter putBlobOperationRate;
+  public final Meter getBlobInfoOperationRate;
+  public final Meter getBlobOperationRate;
+  public final Meter deleteBlobOperationRate;
   public final Meter operationQueuingRate;
   public final Meter operationDequeuingRate;
 
+  // Latency.
   public final Histogram putBlobOperationLatencyMs;
   public final Histogram putChunkOperationLatencyMs;
+  public final Histogram getBlobInfoOperationLatencyMs;
+  public final Histogram getBlobOperationLatencyMs;
+  public final Histogram deleteBlobOperationLatencyMs;
   public final Histogram routerRequestLatencyMs;
 
-  // Operation Errors
-  public final Counter slippedPutSuccessCount;
+  // Operation error count.
   public final Counter putBlobErrorCount;
+  public final Counter getBlobInfoErrorCount;
+  public final Counter getBlobErrorCount;
+  public final Counter deleteBlobErrorCount;
   public final Counter operationAbortCount;
 
-  // Gauge
+  // Count for various errors.
+  public final Counter ambryUnavailableErrorCount;
+  public final Counter invalidBlobIdErrorCount;
+  public final Counter invalidPutArgumentErrorCount;
+  public final Counter operationTimedOutErrorCount;
+  public final Counter routerClosedErrorCount;
+  public final Counter unexpectedInternalErrorCount;
+  public final Counter blobTooLargeErrorCount;
+  public final Counter badInputChannelErrorCount;
+  public final Counter insufficientCapacityErrorCount;
+  public final Counter blobDeletedErrorCount;
+  public final Counter blobDoesNotExistErrorCount;
+  public final Counter blobExpiredErrorCount;
+  public final Counter unknownReplicaResponseError;
+  public final Counter closeErrorCount;
+  public final Counter unknownErrorCountForOperation;
+
+  // Misc metrics.
+  public final Meter operationErrorRate;
+  public final Counter slippedPutSuccessCount;
+  public final Counter ignoredResponseCount;
   public Gauge<Long> chunkFillerThreadRunning;
   public Gauge<Long> requestResponseHandlerThreadRunning;
   public Gauge<Integer> activeOperations;
 
-  public NonBlockingRouterMetrics(MetricRegistry metricRegistry) {
-    this.metricRegistry = metricRegistry;
-    putBlobOperationRate = metricRegistry.meter(MetricRegistry.name(PutManager.class, "-PutBlobRequestArrivalRate"));
-    operationQueuingRate = metricRegistry.meter(MetricRegistry.name(NonBlockingRouter.class, "-OperationQueuingRate"));
+  // Map that stores dataNode-level metrics.
+  private final Map<DataNodeId, NodeLevelMetrics> dataNodeToMetrics;
+
+  private Logger logger = LoggerFactory.getLogger(getClass());
+
+  public NonBlockingRouterMetrics(ClusterMap clusterMap) {
```

Comment:
Because in router metrics there will be per-node metrics. ClusterMap is needed to get all nodes. The frontend metrics has a similar set up.